lib/recurly.d.ts

```diff
@@ -0,0 +1,785 @@
+  shippingAddress?: ShippingAddress;
+  startDate?: Date;
+  endDate?: Date;
+  createdAt?: Date;
+  updatedAt?: Date;
+}
+
+export interface InvoiceCollection {
+  object?: string;
+  chargeInvoice?: Invoice;
+  creditInvoices?: Invoice[];
+}
+
+export interface AccountNote {
+  id?: string;
+  object?: string;
+  accountId?: string;
+  user?: User;
+  message?: string;
+  createdAt?: Date;
+}
+
+export interface User {
+  id?: string;
+  object?: string;
+  email?: string;
+  firstName?: string;
+  lastName?: string;
+  timeZone?: string;
+  createdAt?: Date;
+  deletedAt?: Date;
+}
+
+export interface Subscription {
+  id?: string;
+  object?: string;
+  uuid?: string;
+  account?: AccountMini;
+  plan?: PlanMini;
+  state?: string;
+  shipping?: SubscriptionShipping;
+  couponRedemptions?: CouponRedemptionMini[];
+  pendingChange?: SubscriptionChange;
+  currentPeriodStartedAt?: Date;
+  currentPeriodEndsAt?: Date;
+  currentTermStartedAt?: Date;
+  currentTermEndsAt?: Date;
+  trialStartedAt?: Date;
+  trialEndsAt?: Date;
+  remainingBillingCycles?: number;
+  totalBillingCycles?: number;
+  renewalBillingCycles?: number;
+  autoRenew?: boolean;
+  pausedAt?: Date;
+  remainingPauseCycles?: number;
+  currency?: string;
+  unitAmount?: number;
+  quantity?: number;
+  addOns?: SubscriptionAddOn[];
+  addOnsTotal?: number;
+  subtotal?: number;
+  collectionMethod?: string;
+  poNumber?: string;
+  netTerms?: number;
+  termsAndConditions?: string;
+  customerNotes?: string;
+  expirationReason?: string;
+  customFields?: CustomField[];
+  createdAt?: Date;
+  updatedAt?: Date;
+  activatedAt?: Date;
+  canceledAt?: Date;
+  expiresAt?: Date;
+  bankAccountAuthorizedAt?: Date;
+}
+
+export interface SubscriptionShipping {
+  object?: string;
+  address?: ShippingAddress;
+  method?: ShippingMethodMini;
+  amount?: number;
+}
+
+export interface ShippingMethodMini {
+  id?: string;
+  object?: string;
+  code?: string;
+  name?: string;
+}
+
+export interface CouponRedemptionMini {
+  id?: string;
+  object?: string;
+  coupon?: CouponMini;
+  state?: string;
+  discounted?: number;
+  createdAt?: Date;
+}
+
+export interface CouponMini {
+  id?: string;
+  object?: string;
+  code?: string;
+  name?: string;
+  state?: string;
+  discount?: CouponDiscount;
+  couponType?: string;
+  expiredAt?: Date;
+}
+
+export interface SubscriptionChange {
+  id?: string;
+  object?: string;
+  subscriptionId?: string;
+  plan?: PlanMini;
+  addOns?: SubscriptionAddOn[];
+  unitAmount?: number;
+  quantity?: number;
+  shipping?: SubscriptionShipping;
+  activateAt?: Date;
+  activated?: boolean;
+  createdAt?: Date;
+  updatedAt?: Date;
+  deletedAt?: Date;
+}
+
+export interface SubscriptionAddOn {
+  id?: string;
+  object?: string;
+  subscriptionId?: string;
+  addOn?: AddOnMini;
+  quantity?: number;
+  unitAmount?: number;
+  createdAt?: Date;
+  updatedAt?: Date;
+  expiredAt?: Date;
+}
+
+export interface AddOnMini {
+  id?: string;
```

Comment:
may i ask why all the fields on your objects are optional? i'm honestly not familiar with the API, so is that a property of the API where all fields are optional? like is it similar to facebook's graph API or something where you only request the fields you need?

Reply:
My understanding of the `optional` flag in typescript was that it only had to do with what was optional to construct the object. In theory, there is nothing stopping you from constructing this object without any fields and I thought for the sake of stubbing that it would be best to leave that open. What do you think? Was I interpreting it wrong?

> like is it similar to facebook's graph API or something where you only request the fields you need?

It's not, however we don't yet have the semantics in our OpenAPI spec that describes what must be in a response vs what is optional. That will be coming soon though.

Reply:
It has to do with what's optional to construct the object, but it also means that all of these fields could be undefined. So for example when accessing this id field, it's typed as string|undefined, so you would have to perform a nullability check before doing anything with it.

If there are fields that are guaranteed to be set then it'd be helpful to have them be non optional so that users don't have to perform nullability checks everywhere. It also serves as documentation as to what's available vs not.

Your point about stubbing is a good one though. But in my opinion the documentational aspect of knowing what's available vs not is quite nice. Users of this library can always use the Partial<T> type to make all of an object's fields optional, if they want to have a strongly typed stub.

Reply:
> So for example when accessing this id field, it's typed as string|undefined, so you would have to perform a nullability check before doing anything with it.

Oh boy, I suspected that but was not sure. Like I said before, we don't really have that info in our spec yet. Let me think about how best to handle this. Maybe now is the time to work on requiring that as part of our spec.

One thing I can say for sure, we don't selectively render any json keys. It's either a value or null. So I also don't think `string|undefined` is the right type either. Is there a way to say that a value is `something | null`? Should they just use that union operator? Or is `string | null` the same as just `string` ?

Reply:
Maybe I'm conflating the way typescript handles nulls vs undefineds here.

Reply:
yep `string | null` is a thing and it is not the same as string (typescript will enforce checking truthiness before calling methods on it etc), so that's probably the right type for your stuff.

typescript nulls/undefineds is pretty confusing. they are different, as in javascript, but there's a ton of permutations of ways to define them. `string | undefined` is a distinct type. the question mark syntax, where you put a question mark before the type name, that means it's _optional_ (typed as string|undefined) but the distinction is that you're not required to provide it when constructing the object. if something is explicitly typed as string|undefined then you must provide it as part of the object literal. you can likewise combine these, like `foo?: string | null`. which is typed as `string | null | undefined` but again being optional that means you can omit it in object literal construction.

Reply:
@wcauchois Again, your consultation is super helpful. Thank you!

> the question mark syntax, where you put a question mark before the type name, that means it's optional (typed as string|undefined)

Interesting. According to the rules of our API (specifically that we never selectively define keys), these properties will never be `undefined` in responses. So when constructed from responses at least, I'm confident that `null` is their bottom type. Which I suppose means that the `?` is the right operator to use at all. Perhaps the properties types should be `foo: string` for a required property and `foo: string | null` for an "nullable" property?

>  but the distinction is that you're not required to provide it when constructing the object.

I'm still a little confused as to why these two behaviors are coupled that way. I guess I need to do some more research into what it means to construct objects in typescript. I'm assuming it has compile-type semantics which prevent you from creating "invalid" data and not setting a property leaves it `undefined`. I'm going to need to do some experiments in our test rig to understand this better. I want to make it easy for people to construct these objects when testing, but also it's probably good to optimize for the normal consumption case (in production code not in the test code).